lib/core/test/unit/Cardano/Wallet/Primitive/ScriptsSpec.hs

```diff
@@ -0,0 +1,383 @@
+import qualified Data.ByteString.Char8 as B8
+import qualified Data.List as L
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+spec :: Spec
+spec = do
+    describe "isShared" $ do
+        it "script composed with all of our verification keys should discover them all"
+            (property prop_scriptFromOurVerKeys)
+        it "script composed with not our verification keys should not be discovered"
+            (property prop_scriptFromNotOurVerKeys)
+        it "the same script discovered twice should have the same knownScripts imprint"
+            (property prop_scriptDiscoveredTwice)
+        it "knownScripts of the sequential state is populated properly"
+            (property prop_scriptUpdatesStateProperly)
+        it "scripts with our verification keys are discovered properly"
+            (property prop_scriptsDiscovered)
+        it "discovering our verification keys make them mark Used"
+            (property prop_markingDiscoveredVerKeys)
+        it "discovering works after pool extension"
+            (property prop_poolExtension)
+        it "our verification keys beyond pool gap are not discovered"
+            (property prop_scriptsOutsideNotDiscovered)
+        it "before and after discovery pool number of last consequitive Unused keys stays the same"
+            (property prop_unusedVerKeysConstant)
+        it "discovered verification keys in scripts are consistent between knownScripts and indexedKeys"
+            (property prop_verKeysConsistent)
+
+prop_scriptFromOurVerKeys
+    :: AccountXPubWithScripts
+    -> Property
+prop_scriptFromOurVerKeys (AccountXPubWithScripts accXPub' scripts') = do
+    let (script:_) = scripts'
+    let scriptKeyHashes = retrieveAllVerKeyHashes script
+    let seqState = initializeState accXPub'
+    let (ourSharedKeys, _) = isShared script seqState
+    L.sort (L.nub $ map hashVerificationKey ourSharedKeys) ===
+        L.sort (L.nub scriptKeyHashes)
+
+prop_scriptFromNotOurVerKeys
+    :: ShelleyKey 'AccountK XPub
+    -> AccountXPubWithScripts
+    -> Property
+prop_scriptFromNotOurVerKeys accXPub' (AccountXPubWithScripts _accXPub scripts') = do
+    let (script:_) = scripts'
+    let seqState = initializeState accXPub'
+    let (ourSharedKeys, _) = isShared script seqState
+    ourSharedKeys === []
+
+prop_scriptUpdatesStateProperly
+    :: AccountXPubWithScripts
+    -> Property
+prop_scriptUpdatesStateProperly (AccountXPubWithScripts accXPub' scripts') = do
+    let (script:_) = scripts'
+    let sciptKeyHashes = retrieveAllVerKeyHashes script
+    let seqState = initializeState accXPub'
+    let (_, seqState') = isShared script seqState
+    let scriptKeyHashesInMap =
+            Set.fromList . map (deriveKeyHash accXPub') <$>
+            Map.lookup (toScriptHash script) (getKnownScripts seqState')
+    scriptKeyHashesInMap === Just (Set.fromList (L.nub sciptKeyHashes))
```

Comment:
This property only works if all keys in the script are ours, right? Otherwise, `scriptKeyHashes` may contain script keys that are in the script, but not in `scriptKeyHashesInMap` (since this only tracks our keys in the script).

Reply:
Looking at the generator, I could confirm this :point_up: 

Wouldn't it be better to generate _more realistic_ scripts that can also contain keys that aren't ours? And review this property to say that `scriptKeyHashesInMap` should be included in `scriptKeyHashes`, but not equal.

Reply:
added new property in which two account public keys are engaged. see done in https://github.com/input-output-hk/cardano-wallet/pull/2308/commits/ab8567640737c089cb9f15c4be3cfce6dbcadb1a